### PR TITLE
Support for global hybrids and CAM functionals

### PR DESCRIPTION
**Pull request by vanderhe**
Introduces general infrastructure for global Hybrids and CAM (Coulomb-Attenuating Method) functionals. In particular, four new functionals are available with this PR:

Global Hybrids: B3LYP, PBE0
CAM: CAM-B3LYP, CAM-PBEh

Done/remaining tasks:
- [x] uninitialized floats should be detected more safely due to pre-initialization of reals and integers in debug mode
- [x] compiler toolchain selection corrected
- [x] some non-standard conforming structures converted to standard-conforming ones to compile in F2018 compliance mode
- [x] Poisson solver and finite differences cleaned up
- [x] Hubbard U correction algorithm generalized to CAM-functionals
- [x] check total energy course vs. range-separation parameter (checked against NWCHEM)
- [x] test long-range Hartree-Fock contribution to dimer Hamiltonian for simple limiting case
(reproduces (by approx. 10^(-10)a.u.) long-range Hartree-Fock contribution routine for large range-separation parameters)
- [x] verify reproducibility of old functionals --> largest deviations observed throughout: ~10^(-7)a.u.
(LDA-PW91, GGA-PBE96 checked against old skprogs, GGA-BLYP, LCY-PBE96, LCY-BNL checked against Regensburg)
- more in-depth validation of two-center code for new functionals (no idea how)

### CODE REVIEW

File: common/lib/fdiff.F90 [hunk 1]
```diff
@@ -12,773 +11,264 @@ module common_finitedifferences
...
+  !! On boundaries non-centered five and six point formulae are used.
+  subroutine makeHelmholzFDMatrix7P(M, B, alpha, beta, gama, f0, f1)
+
+    !> finite differences matrix, 7-point scheme
+    real(dp), intent(out), allocatable :: M(:,:)
+
+    !> right-hand side of equation
+    real(dp), intent(inout) :: B(:)
+
+    !>
```
Comment (aradi) [inline common/lib/fdiff.F90:29]: Is the purpose of those arguments not known?

File: slateratom/lib/dft.f90 [hunk 2]
```diff
@@ -279,62 +303,120 @@ subroutine density_grid(pp, max_l, num_alpha, poly_order, alpha, num_mesh_points
...
+        call radial_divergence(tmpv1, abcissa, dz, tmpv2, dzdr)
+        vxc(:, iSpin) = vxc(:, iSpin) - tmpv2
+      end do
+    end select
+
+    !! -------- Exchange + Correlation energy and potential -----------
+
+    select case (xcnr)
+    ! PBE0, B3LYP (global hybrids), CAMY-B3LYP (CAMY-functional)
+    case(7:9)
```
Comment (aradi) [inline slateratom/lib/dft.f90:353]: This we should turn into symbolic enumerations!

**Review by aradi — APPROVED** [review]

File: slateratom/lib/dft.f90 (hunk 2 above)
Comment (vanderhe) [reply slateratom/lib/dft.f90:353]: Agree, I must have overlooked these lines back then.

File: common/lib/fdiff.F90 (hunk 1 above)
Comment (vanderhe) [reply common/lib/fdiff.F90:29]: I tried to understand/comment as much code from the Regensburg-repo as possible, but due to the sheer amount I got sloppy at some point.

**Pull request merged**